src/db/util.rs

```diff
@@ -1,6 +1,105 @@
+#![allow(proc_macro_derive_resolution_fallback)]
+
 use chrono::offset::Utc;
+use diesel::{
+    backend::Backend,
+    deserialize::{self, FromSql},
+    sql_types::BigInt,
+};
+
+use super::{DbError, DbErrorKind};
 
 /// Get the time since the UNIX epoch in milliseconds
 pub fn ms_since_epoch() -> i64 {
     Utc::now().timestamp_millis()
 }
+
+/// Sync Timestamp
+///
+/// Internally represents a Sync timestamp as a u64 representing milliseconds since the epoch.
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Deserialize, Serialize, FromSqlRow)]
+pub struct SyncTimestamp(u64);
+
+impl SyncTimestamp {
+    /// Create a string value compatible with existing Sync Timestamp headers
+    ///
+    /// Represents the timestamp as second since epoch with two decimal places of precision.
+    pub fn as_header(&self) -> String {
+        format!("{:.*}", 2, &self.0)
+    }
+
+    /// Create a `SyncTimestamp` from a string header
+    ///
+    /// Assumes the string represents the seconds since epoch with two decimal places of precision.
+    pub fn from_header(val: &str) -> Result<Self, &'static str> {
+        val.parse::<f64>()
+            .map_err(|_| "Invalid value")
+            .and_then(|v| {
+                if v < 0f64 {
+                    Err("Invalid value")
+                } else {
+                    Ok(v)
+                }
+            }).map(|v: f64| (v * 1_000f64) as u64)
+            .map(SyncTimestamp::from_milliseconds)
+            .map_err(|_| "Invalid value")
+    }
+
+    /// Create a `SyncTimestamp` from an i64
+    ///
+    /// Only called from the db module
+    pub(super) fn from_i64(val: i64) -> Result<Self, DbError> {
```

Comment:
Ah, it's a shame the `TryFrom` trait is nightly-only, it would have been perfect for this right?

Reply:
Yup, I was wishing that as we added it.